src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -1754,7 +1754,28 @@ size_t MergeTreeData::clearOldPartsFromFilesystem(bool force)
     return parts_to_remove.size();
 }
 
-void MergeTreeData::clearPartsFromFilesystem(const DataPartsVector & parts_to_remove)
+void MergeTreeData::clearPartsFromFilesystem(const DataPartsVector & parts)
+{
+    clearPartsFromFilesystemImpl(parts, true, nullptr);
+}
+
+void MergeTreeData::tryClearPartsFromFilesystem(const DataPartsVector & parts, NameSet & parts_failed_to_delete)
+{
+    NameSet part_names_successeded;
+    clearPartsFromFilesystemImpl(parts, false, &part_names_successeded);
+
+    if (part_names_successeded.size() == parts.size())
+        return;
+
+    for (const auto & part : parts)
+    {
+        if (!part_names_successeded.contains(part->name))
+            parts_failed_to_delete.insert(part->name);
+    }
+}
+
+
+void MergeTreeData::clearPartsFromFilesystemImpl(const DataPartsVector & parts_to_remove, bool throw_on_error, NameSet * part_names_successed)
```

Comment:
I think it would be more elegant to pass `std::vector<char> was_part_removed` or something ...
You may avoid adding mutex (but it is also ok).

Reply:
Don't want to do it at this moment.